File: src/main/kotlin/io/instagram/instagram/user/UserController.java
//package io.instagram.instagram.user;
//
//import io.instagram.instagram.auth.JwtUtil;
//import io.instagram.instagram.auth.LoginRequest;
//import io.instagram.instagram.jwt.TokenResponse;
//import lombok.AllArgsConstructor;
//import org.springframework.http.ResponseEntity;
//import org.springframework.security.authentication.AuthenticationManager;
//import org.springframework.security.authentication.BadCredentialsException;
//import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
//import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
//import org.springframework.web.bind.annotation.PostMapping;
//import org.springframework.web.bind.annotation.RequestBody;
//import org.springframework.web.bind.annotation.RestController;
//
//import java.util.Optional;
//
//@AllArgsConstructor
//@RestController
//public class UserController {
//    private final AuthenticationManager authenticationManager;
//    private final AppBaseUserRepository appBaseUserRepository;
//    private final BCryptPasswordEncoder bCryptPasswordEncoder;
//    private final JwtUtil jwtUtil;
//
//    @PostMapping("signup")
//    public AppBaseUser signUp(@RequestBody AppBaseUser user) {
//        user.setPassword(bCryptPasswordEncoder.encode(user.getPassword()));
//        return appBaseUserRepository.save(user);
//    }
//
//    @PostMapping("login")
//    public ResponseEntity<TokenResponse> loginUser(@RequestBody
//                                                   LoginRequest loginRequest) throws BadCredentialsException {
//        try {
//            authenticationManager.authenticate(
//                    new UsernamePasswordAuthenticationToken(loginRequest.getUsername(),
//                                                            loginRequest.getPassword()));
//        } catch (BadCredentialsException e) {
//            throw new BadCredentialsException("Username or Password is incorrect");
//        }
//
//        final Optional<AppBaseUser> user = appBaseUserRepository.findByUsername(loginRequest.getUsername());
//
//        if (user.isPresent()) {
//            final TokenResponse tokenResponse = jwtUtil.generateTokens(user.get());
//            return ResponseEntity.ok(tokenResponse);
//        } else {
//            throw new BadCredentialsException("Username or Password is incorrect");
//        }
//    }
//}

File: src/main/kotlin/io/instagram/instagram/user/UserController.kt
package io.instagram.instagram.user

import io.instagram.instagram.auth.JwtUtil
import io.instagram.instagram.auth.LoginRequest
import io.instagram.instagram.jwt.TokenResponse
import org.springframework.http.ResponseEntity
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.authentication.InternalAuthenticationServiceException
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController

@RestController
class UserController(private val authenticationManager: AuthenticationManager,
                     private val jwtUtil: JwtUtil,
                     private val bCryptPasswordEncoder: BCryptPasswordEncoder,
                     private val repository: UserRepository) {

    @PostMapping("signup")
    fun signUp(@RequestBody user: AppBaseUser): AppBaseUser {
        user.password = bCryptPasswordEncoder.encode(user.password)
        return repository.save(user)
    }

    @PostMapping("login")
    @Throws(BadCredentialsException::class,
            InternalAuthenticationServiceException::class)
    fun loginUser(@RequestBody loginRequest: LoginRequest): ResponseEntity<TokenResponse> {
        try {
            authenticationManager.authenticate(UsernamePasswordAuthenticationToken(loginRequest.username,
                    loginRequest.password))
        } catch (e: InternalAuthenticationServiceException) {
            throw InternalAuthenticationServiceException("No such user with this username.")
        } catch (e: BadCredentialsException) {
            throw BadCredentialsException("Username or Password is incorrect")
        }
        val user: AppBaseUser = repository.findByUsername(loginRequest.username)
        return ResponseEntity.ok(jwtUtil.generateAuthToken(user))
    }

    @GetMapping("users")
    fun getUsers(): MutableList<AppBaseUser> {
        return repository.findAll()
    }
}